content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+   	mp = (unreadable could not find loclist entry at 0x2f45f for address 0x42f32b)
+   	status = (unreadable could not find loclist entry at 0x2f4c6 for address 0x42f32b)
+
+1  0x000000000042f3d3 in runtime.goparkunlock
+   at /usr/local/go/src/runtime/proc.go:308
+   	lock = (unreadable empty OP stack)
+   	reason = (unreadable empty OP stack)
+   	traceEv = (unreadable empty OP stack)
+   	traceskip = (unreadable empty OP stack)
+
+2  0x00000000004069cd in runtime.chansend
+   at /usr/local/go/src/runtime/chan.go:234
+   	block = true
+   	c = (*runtime.hchan)(0xc000082480)
+   	callerpc = (unreadable empty OP stack)
+   	ep = unsafe.Pointer(0xc000031e30)
+   	~r4 = (unreadable empty OP stack)
+   	gp = (*runtime.g)(0xc000001680)
+   	mysg = *(unreadable read out of bounds)
+   	t0 = 0
+
+3  0x00000000004067a5 in runtime.chansend1
+   at /usr/local/go/src/runtime/chan.go:125
+   	c = (unreadable empty OP stack)
+   	elem = (unreadable empty OP stack)
+
+4  0x0000000000703b8c in main.(*Metrics).SetM
+   at /vagrant/example/server/metrics.go:145
+   	key = "metrics.raw_channel"
+   	m = (*main.Metrics)(0xc000134000)
+   	value = 100
+
+5  0x0000000000704394 in main.(*Metrics).sendMetricsToOutChannel
+   at /vagrant/example/server/metrics.go:206
+   	m = (*main.Metrics)(0xc000134000)
+   	scope = 0
+   	updateInterval = (unreadable could not find loclist entry at 0x9d4ed for address 0x704393)
+
+6  0x0000000000703f40 in main.(*Metrics).startInChannelConsumer
+   at /vagrant/example/server/metrics.go:184
+   	m = (*main.Metrics)(0xc000134000)
+   	inMetrics = main.Metric {Type: TypeCount, Scope: 0, Key: "server.req-incoming",...+2 more}
+   	nextUpdate = (unreadable could not find loclist entry at 0x9d3fd for address 0x703f3f)
+
+7  0x000000000045cf11 in runtime.goexit
+   at /usr/local/go/src/runtime/asm_amd64.s:1333
+```
+
+Bottom to top:
+
+(6) At `(*Metrics).startInChannelConsumer:184` a new value from the channel has been received
+
+(5) We called `(*Metrics).sendMetricsToOutChannel` first
+
+(4) And `main.(*Metrics).SetM` next, passing `key="metrics.raw_channel"` and `value = 100`
+
+And so on until we've been blocked in `runtime.gopark` with “waitReasonChanSend”.
+
+Everything makes sense now!
+
+Within a single goroutine, the function that reads values out of a buffered channel tried to put additional values into the channel. As the number of incoming values to the channel became close to its capacity, the consumer-function deadlocked itself trying to add value to the full channel. Since the single channel's consumer was deadlocked, every new incoming request that tried adding values into the channel became blocked as well.
+
+----
+
+And that’s our story. Using the described technique we’ve managed to find the root-cause the problem. The original piece of code was written many years ago. Nobody even looked at it and never thought it might bring such issues.
```

Comment:
```suggestion
And that’s our story. Using the described technique we’ve managed to find the root cause of the problem. The original piece of code was written many years ago. Nobody even looked at it and never thought it might bring such issues.
```